modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+     *
+     * @param opCtx Snapshot restore operation context.
+     */
+    private void rollback(@Nullable SnapshotRestoreContext opCtx) {
+        if (opCtx == null || F.isEmpty(opCtx.dirs))
+            return;
+
+        if (log.isInfoEnabled())
+            log.info("Performing local rollback routine for restored cache groups [requestID=" + opCtx.reqId + ']');
+
+        try {
+            for (File cacheDir : opCtx.dirs) {
+                if (!cacheDir.exists())
+                    continue;
+
+                if (log.isInfoEnabled())
+                    log.info("Cleaning up directory " + cacheDir);
+
+                U.delete(cacheDir);
+            }
+        }
+        catch (Exception e) {
+            log.error("Failed to perform rollback [requestID=" + opCtx.reqId + ", snapshot=" + opCtx.snpName + ']', e);
+        }
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishPrepare(UUID reqId, Map<UUID, ArrayList<StoredCacheData>> res, Map<UUID, Exception> errs) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (ctx.clientNode())
+            return;
+
+        Exception failure = checkFailure(errs, opCtx0, res.keySet());
+
+        if (failure == null) {
+            assert opCtx0 != null : ctx.localNodeId();
+
+            Map<Integer, StoredCacheData> globalCfgs = new HashMap<>();
+
+            for (List<StoredCacheData> storedCfgs : res.values()) {
+                if (storedCfgs == null)
+                    continue;
+
+                for (StoredCacheData cacheData : storedCfgs)
+                    globalCfgs.put(CU.cacheId(cacheData.config().getName()), cacheData);
+            }
+
+            opCtx0.cfgs = globalCfgs;
+
+            if (U.isLocalNodeCoordinator(ctx.discovery()))
+                cacheStartProc.start(reqId, reqId);
+
+            return;
+        }
+
+        if (opCtx0 == null)
+            finishProcess(failure);
+        else // Remove files asynchronously.
+            ctx.cache().context().snapshotMgr().snapshotExecutorService().execute(() -> {
+                rollback(opCtx0);
+
+                finishProcess(failure);
+            });
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @return Result future.
+     */
+    private IgniteInternalFuture<Boolean> cacheStart(UUID reqId) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 == null)
+            return new GridFinishedFuture<>();
+
+        if (!reqId.equals(opCtx0.reqId)) {
+            return new GridFinishedFuture<>(
+                new IgniteCheckedException("Unknown snapshot restore operation was rejected."));
+        }
+
+        if (!U.isLocalNodeCoordinator(ctx.discovery()))
+            return new GridFinishedFuture<>();
+
+        DiscoveryDataClusterState state = ctx.state().clusterState();
+
+        if (state.state() != ClusterState.ACTIVE || state.transition())
```

Comment:
I think you should not check the sate of the cluster each time you run the next stage. If the state is changed it might be better to set the correct exception for the `opCtx.err` field and check it each time you want to proceed.

Reply:
How to handle the transition from/to `read-only` mode?